File: backend/src/services/whatsappService.js
import axios from 'axios';
import dotenv from 'dotenv';

dotenv.config();

const WHATSAPP_API_URL = process.env.WHATSAPP_API_URL;

// Helper to get headers for a specific store
const getHeaders = (accessToken) => ({
    'Authorization': `Bearer ${accessToken || process.env.WHATSAPP_ACCESS_TOKEN}`,
    'Content-Type': 'application/json'
});

// Helper to get URL for a specific store
const getUrl = (phoneNumberId) => {
    const id = phoneNumberId || process.env.WHATSAPP_PHONE_NUMBER_ID;
    return `${WHATSAPP_API_URL}/${id}/messages`;
};

export const sendTextMessage = async (to, message, storeConfig = {}) => {
    try {
        const response = await axios.post(
            getUrl(storeConfig.phoneNumberId),
            {
                messaging_product: 'whatsapp',
                to: to,
                type: 'text',
                text: { body: message }
            },
            { headers: getHeaders(storeConfig.accessToken) }
        );
        return response.data;
    } catch (error) {
        console.error('Error sending WhatsApp message:', error.response?.data || error.message);
        throw error;
    }
};

export const sendInteractiveButtons = async (to, bodyText, buttons, storeConfig = {}) => {
    try {
        const response = await axios.post(
            getUrl(storeConfig.phoneNumberId),
            {
                messaging_product: 'whatsapp',
                to: to,
                type: 'interactive',
                interactive: {
                    type: 'button',
                    body: { text: bodyText },
                    action: {
                        buttons: buttons.map((btn, idx) => ({
                            type: 'reply',
                            reply: {
                                id: btn.id || `btn_${idx}`,
                                title: btn.title.substring(0, 20) // Max 20 chars
                            }
                        }))
                    }
                }
            },
            { headers: getHeaders(storeConfig.accessToken) }
        );
        return response.data;
    } catch (error) {
        console.error('Error sending interactive buttons:', error.response?.data || error.message);
        throw error;
    }
};

export const sendInteractiveList = async (to, bodyText, buttonText, sections, storeConfig = {}) => {
    try {
        const response = await axios.post(
            getUrl(storeConfig.phoneNumberId),
            {
                messaging_product: 'whatsapp',
                to: to,
                type: 'interactive',
                interactive: {
                    type: 'list',
                    body: { text: bodyText },
                    action: {
                        button: buttonText,
                        sections: sections
                    }
                }
            },
            { headers: getHeaders(storeConfig.accessToken) }
        );
        return response.data;
    } catch (error) {
        console.error('Error sending interactive list:', error.response?.data || error.message);
        throw error;
    }
};

export const sendImageMessage = async (to, imageUrl, caption = '', storeConfig = {}) => {
    try {
        const response = await axios.post(
            getUrl(storeConfig.phoneNumberId),
            {
                messaging_product: 'whatsapp',
                to: to,
                type: 'image',
                image: {
                    link: imageUrl,
                    caption: caption
                }
            },
            { headers: getHeaders(storeConfig.accessToken) }
        );
        return response.data;
    } catch (error) {
        console.error('Error sending image:', error.response?.data || error.message);
        throw error;
    }
};

export const sendTemplateMessage = async (to, templateName, languageCode = 'en', storeConfig = {}) => {
    try {
        const response = await axios.post(
            getUrl(storeConfig.phoneNumberId),
            {
                messaging_product: 'whatsapp',
                to: to,
                type: 'template',
                template: {
                    name: templateName,
                    language: { code: languageCode }
                }
            },
            { headers: getHeaders(storeConfig.accessToken) }
        );
        return response.data;
    } catch (error) {
        console.error('Error sending template:', error.response?.data || error.message);
        throw error;
    }
};

export const markAsRead = async (messageId, storeConfig = {}) => {
    try {
        await axios.post(
            getUrl(storeConfig.phoneNumberId),
            {
                messaging_product: 'whatsapp',
                status: 'read',
                message_id: messageId
            },
            { headers: getHeaders(storeConfig.accessToken) }
        );
    } catch (error) {
        console.error('Error marking message as read:', error.response?.data || error.message);
    }
};

File: backend-dotnet/Wazflo.Api/Services/WhatsAppService.cs
using System.Net.Http.Headers;
using System.Text;
using Newtonsoft.Json;

namespace Wazflo.Api.Services
{
    public class WhatsAppService : IWhatsAppService
    {
        private readonly HttpClient _httpClient;
        private readonly IConfiguration _configuration;
        private readonly string _baseUrl;

        public WhatsAppService(HttpClient httpClient, IConfiguration configuration)
        {
            _httpClient = httpClient;
            _configuration = configuration;
            _baseUrl = _configuration["WhatsApp:ApiUrl"] ?? "https://graph.facebook.com/v17.0";
        }

        public async Task<object> SendTextMessageAsync(string to, string message, string phoneNumberId, string accessToken)
        {
            var payload = new
            {
                messaging_product = "whatsapp",
                to = to,
                type = "text",
                text = new { body = message }
            };

            return await PostToWhatsAppAsync(phoneNumberId, accessToken, payload);
        }

        public async Task<object> SendInteractiveButtonsAsync(string to, string bodyText, List<WhatsAppButton> buttons, string phoneNumberId, string accessToken)
        {
            var payload = new
            {
                messaging_product = "whatsapp",
                to = to,
                type = "interactive",
                interactive = new
                {
                    type = "button",
                    body = new { text = bodyText },
                    action = new
                    {
                        buttons = buttons.ConvertAll(btn => new
                        {
                            type = "reply",
                            reply = new
                            {
                                id = btn.Id,
                                title = btn.Title.Length > 20 ? btn.Title.Substring(0, 20) : btn.Title
                            }
                        })
                    }
                }
            };

            return await PostToWhatsAppAsync(phoneNumberId, accessToken, payload);
        }

        public async Task<object> SendImageMessageAsync(string to, string imageUrl, string caption, string phoneNumberId, string accessToken)
        {
            var payload = new
            {
                messaging_product = "whatsapp",
                to = to,
                type = "image",
                image = new
                {
                    link = imageUrl,
                    caption = caption
                }
            };

            return await PostToWhatsAppAsync(phoneNumberId, accessToken, payload);
        }

        public async Task<object> SendTemplateMessageAsync(string to, string templateName, string languageCode, string phoneNumberId, string accessToken)
        {
            var payload = new
            {
                messaging_product = "whatsapp",
                to = to,
                type = "template",
                template = new
                {
                    name = templateName,
                    language = new { code = languageCode }
                }
            };

            return await PostToWhatsAppAsync(phoneNumberId, accessToken, payload);
        }

        private async Task<object> PostToWhatsAppAsync(string phoneNumberId, string accessToken, object payload)
        {
            var url = $"{_baseUrl}/{phoneNumberId}/messages";
            var request = new HttpRequestMessage(HttpMethod.Post, url);
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
            request.Content = new StringContent(JsonConvert.SerializeObject(payload), Encoding.UTF8, "application/json");

            var response = await _httpClient.SendAsync(request);
            var content = await response.Content.ReadAsStringAsync();

            if (!response.IsSuccessStatusCode)
            {
                throw new Exception($"WhatsApp API error: {content}");
            }

            return JsonConvert.DeserializeObject<object>(content)!;
        }
    }
}
